pycbc/events/coinc_rate.py

```diff
@@ -0,0 +1,154 @@
+    and signal coincidences.
+"""
+
+import itertools
+import numpy
+import pycbc.detector
+
+
+def multiifo_noise_coinc_rate(rates, slop):
+    """
+    Calculate the expected rate of noise coincidences for multiple detectors
+
+    Parameters
+    ----------
+    rates: dict
+        Dictionary keyed on interferometer
+        Value is a vector of single-detector trigger rates
+    slop: float
+        time to be added to time-of-flights between detectors to account
+        for timing error
+
+    Returns
+    -------
+    expected_coinc_rates: dict
+        A dictionary of ifo combined rates keyed on the combination string
+        value is the expected coincidence rate in the combination given the
+        individual detector trigger rates
+    """
+    ifos = numpy.array(sorted(rates.keys()))
+    rates_raw = list(rates[ifo] for ifo in ifos)
+    expected_coinc_rates = {}
+    # Calculate coincidence for all-ifo combination
+    # multiply the product of rates and by the overlap time
+    allowed_area = multiifo_noise_coincident_area(ifos, slop)
+
+    rateprod = [numpy.prod(rs) for rs in zip(*rates_raw)]
+    ifostring = ' '.join(ifos)
+    expected_coinc_rates[ifostring] = allowed_area * numpy.array(rateprod)
+    # if more than one possible coincidence type exists,
+    # calculate coincidences for subsets through recursion
+    if len(ifos) > 2:
+        # Calculate rate for each 'miss-one-out' detector combination
+        subsets = itertools.combinations(ifos, len(ifos) - 1)
+        for subset in subsets:
+            rates_subset = {}
+            for ifo in subset:
+                rates_subset[ifo] = rates[ifo]
+            sub_coinc_rates = multiifo_noise_coinc_rate(rates_subset, slop)
+            # add these sub-coincidences to the overall dictionary
+            for sub_coinc in sub_coinc_rates:
+                expected_coinc_rates[sub_coinc] = sub_coinc_rates[sub_coinc]
+
+    return expected_coinc_rates
+
+
+def multiifo_noise_coincident_area(ifos, slop):
+    """
+    calculate the total extent of time offset between 2 detectors,
+    or area of the 2d space of time offsets for 3 detectors, for
+    which a coincidence can be generated
+
+    Parameters
+    ----------
+    ifos: list of strings
+        list of interferometers for area calculation
+    slop: float
+        extra time to add on to time-of-flight for timing error
+
+    Returns
+    -------
+    allowed_area: float
+        area in units of seconds^(n_ifos-1) that the coincident values
+        can fall in
+    """
+    dets = {}
+    for ifo in ifos:
+        dets[ifo] = pycbc.detector.Detector(ifo)
+    n_ifos = len(ifos)
+    if n_ifos == 2:
+        allowed_area = 2 * (dets[ifos[0]].light_travel_time_to_detector(
+                                                        dets[ifos[1]]) + slop)
+    elif n_ifos == 3:
+        tofs = numpy.zeros(n_ifos)
+        ifo2_num = []
+        # set up detector objects
```

Comment:
did some code get eaten at this point?

Reply:
I moved the creation of dets (and deprecated det1 and det0) to created outside of the if statement, which meant that the 2-ifo case could become a single line

Reply:
ok, so fix the comment